File: control/pedotransfer/saxton_rawls.py
#!/usr/bin/env python3
"""
Exp 023: Pedotransfer Functions (Saxton & Rawls 2006) — Python Control Baseline

Implements the Saxton & Rawls (2006) pedotransfer functions to estimate soil
hydraulic properties from texture (sand/clay percentages) and organic matter.

Reference:
    Saxton KE, Rawls WJ (2006) "Soil water characteristic estimates by texture
    and organic matter for hydrologic solutions." Soil Sci. Soc. Am. J.
    70(5): 1569-1578.

Equations estimate:
    - Wilting point (θ_wp at -1500 kPa)
    - Field capacity (θ_fc at -33 kPa)
    - Saturation (θ_s = porosity)
    - Saturated hydraulic conductivity (Ksat)

Open data: Uses USDA soil texture class definitions + published regression
    coefficients from Saxton & Rawls (2006).
"""

import json
import math
import os
import sys

# Saxton & Rawls (2006) regression coefficients
# θ_1500 (wilting point) at -1500 kPa
def theta_1500_first(S, C, OM):
    """First estimate of wilting point moisture."""
    return (-0.024 * S + 0.487 * C + 0.006 * OM
            + 0.005 * S * OM - 0.013 * C * OM
            + 0.068 * S * C + 0.031)

def theta_1500(S, C, OM):
    """Wilting point (θ_wp at -1500 kPa)."""
    t1500_first = theta_1500_first(S, C, OM)
    return t1500_first + (0.14 * t1500_first - 0.02)

# θ_33 (field capacity) at -33 kPa
def theta_33_first(S, C, OM):
    """First estimate of field capacity moisture."""
    return (-0.251 * S + 0.195 * C + 0.011 * OM
            + 0.006 * S * OM - 0.027 * C * OM
            + 0.452 * S * C + 0.299)

def theta_33(S, C, OM):
    """Field capacity (θ_fc at -33 kPa)."""
    t33_first = theta_33_first(S, C, OM)
    return t33_first + (1.283 * t33_first * t33_first - 0.374 * t33_first - 0.015)

# θ_s (saturation = porosity) and θ_33_s (saturated moisture at -33 kPa)
def theta_s_33_first(S, C, OM):
    """First estimate of moisture between saturation and field capacity."""
    t33 = theta_33(S, C, OM)
    return (0.278 * S + 0.034 * C + 0.022 * OM
            - 0.018 * S * OM - 0.027 * C * OM
            - 0.584 * S * C + 0.078)

def theta_s_33(S, C, OM):
    """Moisture between saturation and field capacity."""
    first = theta_s_33_first(S, C, OM)
    return first + (0.636 * first - 0.107)

def theta_s(S, C, OM):
    """Saturation moisture (porosity)."""
    return theta_33(S, C, OM) + theta_s_33(S, C, OM) - 0.097 * S + 0.043

# Saturated hydraulic conductivity (Ksat)
def lambda_param(S, C, OM):
    """Slope of ln(moisture) vs ln(tension) curve."""
    t33 = theta_33(S, C, OM)
    t1500 = theta_1500(S, C, OM)
    B = (math.log(1500) - math.log(33)) / (math.log(t33) - math.log(t1500))
    return 1.0 / B

def ksat(S, C, OM):
    """Saturated hydraulic conductivity (mm/hr)."""
    ts = theta_s(S, C, OM)
    t33 = theta_33(S, C, OM)
    lam = lambda_param(S, C, OM)
    return 1930.0 * (ts - t33) ** (3.0 - lam)


# USDA texture class definitions (representative sand/clay/OM)
# sand, clay fractions as proportions (0-1); OM as percentage
USDA_TEXTURES = {
    "sand":        {"S": 0.92, "C": 0.03, "OM": 1.0},
    "loamy_sand":  {"S": 0.82, "C": 0.06, "OM": 1.5},
    "sandy_loam":  {"S": 0.65, "C": 0.10, "OM": 2.0},
    "loam":        {"S": 0.40, "C": 0.20, "OM": 2.5},
    "silt_loam":   {"S": 0.20, "C": 0.15, "OM": 3.0},
    "clay_loam":   {"S": 0.30, "C": 0.35, "OM": 2.0},
    "silty_clay":  {"S": 0.07, "C": 0.45, "OM": 2.0},
    "clay":        {"S": 0.20, "C": 0.55, "OM": 2.0},
}

# Expected ranges from Saxton & Rawls (2006) Table 4
EXPECTED_RANGES = {
    "theta_wp": (0.01, 0.35),    # volumetric, 0-1
    "theta_fc": (0.05, 0.50),    # volumetric, 0-1
    "theta_s":  (0.30, 0.65),    # porosity
    "ksat":     (0.1, 500.0),    # mm/hr
}


def run_validation():
    """Run all Saxton-Rawls pedotransfer validation checks."""
    passed = 0
    failed = 0

    def check(name, condition, detail=""):
        nonlocal passed, failed
        status = "PASS" if condition else "FAIL"
        if not condition:
            failed += 1
            print(f"  [{status}] {name}: {detail}")
        else:
            passed += 1
            print(f"  [{status}] {name}")

    print("=" * 60)
    print("  Exp 023: Pedotransfer (Saxton-Rawls 2006) Validation")
    print("=" * 60)

    # --- Section 1: Analytical tests ---
    print("\n--- Analytical (Loam: S=0.40, C=0.20, OM=2.5%) ---")
    S, C, OM = 0.40, 0.20, 2.5

    wp = theta_1500(S, C, OM)
    fc = theta_33(S, C, OM)
    sat = theta_s(S, C, OM)
    ks = ksat(S, C, OM)

    check("loam_wp_range", 0.08 < wp < 0.20,
          f"θ_wp={wp:.4f}")
    check("loam_fc_range", 0.20 < fc < 0.40,
          f"θ_fc={fc:.4f}")
    check("loam_sat_range", 0.35 < sat < 0.55,
          f"θ_s={sat:.4f}")
    check("loam_ksat_range", 1.0 < ks < 100.0,
          f"Ksat={ks:.2f} mm/hr")

    # Physical constraints
    check("wp_lt_fc", wp < fc, f"θ_wp={wp:.4f} < θ_fc={fc:.4f}")
    check("fc_lt_sat", fc < sat, f"θ_fc={fc:.4f} < θ_s={sat:.4f}")
    check("ksat_positive", ks > 0.0, f"Ksat={ks:.2f}")

    # --- Section 2: Texture class sweep ---
    print("\n--- All USDA Texture Classes ---")

    results = {}
    for tex_name, tex in USDA_TEXTURES.items():
        S, C, OM = tex["S"], tex["C"], tex["OM"]
        wp = theta_1500(S, C, OM)
        fc = theta_33(S, C, OM)
        sat = theta_s(S, C, OM)
        ks = ksat(S, C, OM)
        results[tex_name] = {"wp": wp, "fc": fc, "sat": sat, "ksat": ks}

        # Physical ordering
        check(f"{tex_name}: wp<fc<sat",
              wp < fc < sat,
              f"wp={wp:.3f} fc={fc:.3f} sat={sat:.3f}")

        # Range checks
        check(f"{tex_name}: wp in range",
              EXPECTED_RANGES["theta_wp"][0] < wp < EXPECTED_RANGES["theta_wp"][1],
              f"θ_wp={wp:.4f}")
        check(f"{tex_name}: fc in range",
              EXPECTED_RANGES["theta_fc"][0] < fc < EXPECTED_RANGES["theta_fc"][1],
              f"θ_fc={fc:.4f}")
        check(f"{tex_name}: sat in range",
              EXPECTED_RANGES["theta_s"][0] < sat < EXPECTED_RANGES["theta_s"][1],
              f"θ_s={sat:.4f}")
        check(f"{tex_name}: ksat in range",
              EXPECTED_RANGES["ksat"][0] < ks < EXPECTED_RANGES["ksat"][1],
              f"Ksat={ks:.2f}")

    # --- Section 3: Ordering checks ---
    print("\n--- Physical Ordering Across Textures ---")

    # Sand should have higher Ksat than clay
    check("sand_ksat_gt_clay",
          results["sand"]["ksat"] > results["clay"]["ksat"],
          f"sand={results['sand']['ksat']:.1f} clay={results['clay']['ksat']:.1f}")

    # Clay should have higher WP than sand
    check("clay_wp_gt_sand",
          results["clay"]["wp"] > results["sand"]["wp"],
          f"clay={results['clay']['wp']:.3f} sand={results['sand']['wp']:.3f}")

    # Clay should have higher FC than sand
    check("clay_fc_gt_sand",
          results["clay"]["fc"] > results["sand"]["fc"],
          f"clay={results['clay']['fc']:.3f} sand={results['sand']['fc']:.3f}")

    # --- Section 4: Organic matter sensitivity ---
    print("\n--- Organic Matter Sensitivity ---")

    S, C = 0.40, 0.20
    wp_lo = theta_1500(S, C, 0.5)
    fc_lo = theta_33(S, C, 0.5)
    wp_hi = theta_1500(S, C, 5.0)
    fc_hi = theta_33(S, C, 5.0)

    # Higher OM → higher water retention
    check("om_increases_wp", wp_hi > wp_lo,
          f"OM=0.5: wp={wp_lo:.4f}, OM=5.0: wp={wp_hi:.4f}")
    check("om_increases_fc", fc_hi > fc_lo,
          f"OM=0.5: fc={fc_lo:.4f}, OM=5.0: fc={fc_hi:.4f}")

    # --- Section 5: Available water capacity ---
    print("\n--- Available Water Capacity ---")

    for tex_name, r in results.items():
        awc = r["fc"] - r["wp"]
        check(f"{tex_name}: AWC positive",
              awc > 0.0,
              f"AWC={awc:.4f}")
        check(f"{tex_name}: AWC realistic",
              0.01 < awc < 0.25,
              f"AWC={awc:.4f}")

    # Loam/silt_loam should have highest AWC (known soil science)
    loam_awc = results["loam"]["fc"] - results["loam"]["wp"]
    sand_awc = results["sand"]["fc"] - results["sand"]["wp"]
    check("loam_awc_gt_sand", loam_awc > sand_awc,
          f"loam={loam_awc:.3f} sand={sand_awc:.3f}")

    # --- Section 6: TAW computation ---
    print("\n--- Total Available Water (TAW) ---")

    root_depth_mm = 900.0  # corn
    loam_taw = (results["loam"]["fc"] - results["loam"]["wp"]) * root_depth_mm
    check("loam_taw_realistic",
          50.0 < loam_taw < 250.0,
          f"TAW={loam_taw:.1f} mm for 900mm root depth")

    # --- Summary ---
    print(f"\n{'=' * 60}")
    total = passed + failed
    print(f"  Pedotransfer (Saxton-Rawls): {passed}/{total} PASS, {failed}/{total} FAIL")
    print(f"{'=' * 60}")

    return failed == 0


def generate_benchmark():
    """Generate benchmark JSON for Rust validation."""
    all_results = {}
    for tex_name, tex in USDA_TEXTURES.items():
        S, C, OM = tex["S"], tex["C"], tex["OM"]
        wp = theta_1500(S, C, OM)
        fc = theta_33(S, C, OM)
        sat = theta_s(S, C, OM)
        ks = ksat(S, C, OM)
        lam = lambda_param(S, C, OM)
        all_results[tex_name] = {
            "S": S, "C": C, "OM": OM,
            "theta_wp": round(wp, 6),
            "theta_fc": round(fc, 6),
            "theta_s": round(sat, 6),
            "ksat_mm_hr": round(ks, 6),
            "lambda": round(lam, 6),
        }

    # Loam analytical intermediate values
    S, C, OM = 0.40, 0.20, 2.5
    loam_intermediates = {
        "S": S, "C": C, "OM": OM,
        "theta_1500_first": round(theta_1500_first(S, C, OM), 6),
        "theta_1500": round(theta_1500(S, C, OM), 6),
        "theta_33_first": round(theta_33_first(S, C, OM), 6),
        "theta_33": round(theta_33(S, C, OM), 6),
        "theta_s_33_first": round(theta_s_33_first(S, C, OM), 6),
        "theta_s_33": round(theta_s_33(S, C, OM), 6),
        "theta_s": round(theta_s(S, C, OM), 6),
        "lambda": round(lambda_param(S, C, OM), 6),
        "ksat_mm_hr": round(ksat(S, C, OM), 6),
    }

    benchmark = {
        "_provenance": {
            "method": "Saxton & Rawls (2006) pedotransfer functions",
            "baseline_script": "control/pedotransfer/saxton_rawls.py",
            "baseline_commit": "pending",
            "python_version": f"{sys.version_info.major}.{sys.version_info.minor}",
            "created": "2026-02-26",
            "references": [
                "Saxton KE, Rawls WJ (2006) SSSAJ 70(5):1569-1578",
            ],
        },
        "loam_intermediates": loam_intermediates,
        "texture_classes": all_results,
        "tol_moisture": 1e-4,
        "tol_ksat": 0.5,
        "expected_ranges": EXPECTED_RANGES,
        "thresholds": {
            "awc_range": [0.01, 0.25],
            "_tolerance_justification": (
                "Saxton-Rawls regression coefficients have 4 significant digits; "
                "intermediate values may accumulate floating-point differences. "
                "Ksat tolerance is wider due to exponential amplification."
            ),
        },
    }

    out_path = os.path.join(os.path.dirname(__file__), "benchmark_pedotransfer.json")
    with open(out_path, "w") as f:
        json.dump(benchmark, f, indent=2)
    print(f"\n  Benchmark written to {out_path}")

    return benchmark


if __name__ == "__main__":
    benchmark = generate_benchmark()
    success = run_validation()
    sys.exit(0 if success else 1)

File: barracuda/src/eco/soil_moisture/saxton_rawls.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Saxton & Rawls (2006) pedotransfer functions.
//!
//! Estimates wilting point, field capacity, saturation, and Ksat from
//! sand fraction, clay fraction, and organic matter percentage.
//!
//! Saxton KE, Rawls WJ (2006) Soil Sci. Soc. Am. J. 70(5):1569-1578.

/// Continuous soil hydraulic properties from Saxton & Rawls (2006) regressions.
#[derive(Debug, Clone, Copy)]
pub struct SaxtonRawlsInput {
    /// Sand fraction (0–1).
    pub sand: f64,
    /// Clay fraction (0–1).
    pub clay: f64,
    /// Organic matter percentage (e.g. 2.5 for 2.5%).
    pub om_pct: f64,
}

/// Result from Saxton-Rawls pedotransfer regression.
#[derive(Debug, Clone, Copy)]
pub struct SaxtonRawlsResult {
    /// Wilting point θ at −1500 kPa (m³/m³).
    pub theta_wp: f64,
    /// Field capacity θ at −33 kPa (m³/m³).
    pub theta_fc: f64,
    /// Saturation moisture content / porosity (m³/m³).
    pub theta_s: f64,
    /// Saturated hydraulic conductivity (mm/hr).
    pub ksat_mm_hr: f64,
    /// Slope parameter λ of the moisture-tension curve.
    pub lambda: f64,
}

// Saxton-Rawls regressions use plain arithmetic to match Python baseline bit-for-bit.
// mul_add reordering changes float rounding enough to break validation tolerances.
#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_1500_first(s: f64, c: f64, om: f64) -> f64 {
    -0.024 * s + 0.487 * c + 0.006 * om + 0.005 * s * om - 0.013 * c * om + 0.068 * s * c + 0.031
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_1500(s: f64, c: f64, om: f64) -> f64 {
    let first = sr_theta_1500_first(s, c, om);
    first + 0.14 * first - 0.02
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_33_first(s: f64, c: f64, om: f64) -> f64 {
    -0.251 * s + 0.195 * c + 0.011 * om + 0.006 * s * om - 0.027 * c * om + 0.452 * s * c + 0.299
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_33(s: f64, c: f64, om: f64) -> f64 {
    let first = sr_theta_33_first(s, c, om);
    first + 1.283 * first * first - 0.374 * first - 0.015
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_s_33_first(s: f64, c: f64, om: f64) -> f64 {
    0.278 * s + 0.034 * c + 0.022 * om - 0.018 * s * om - 0.027 * c * om - 0.584 * s * c + 0.078
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_s_33(s: f64, c: f64, om: f64) -> f64 {
    let first = sr_theta_s_33_first(s, c, om);
    first + 0.636 * first - 0.107
}

#[expect(
    clippy::suboptimal_flops,
    reason = "match reference implementation bit-for-bit"
)]
fn sr_theta_s(s: f64, c: f64, om: f64) -> f64 {
    sr_theta_33(s, c, om) + sr_theta_s_33(s, c, om) - 0.097 * s + 0.043
}

const SR_PRESSURE_1500_KPA: f64 = 1500.0;
const SR_PRESSURE_33_KPA: f64 = 33.0;
const SR_KSAT_COEFF: f64 = 1930.0;
const SR_KSAT_EXPONENT: f64 = 3.0;

fn sr_lambda(s: f64, c: f64, om: f64) -> f64 {
    let t33 = sr_theta_33(s, c, om);
    let t1500 = sr_theta_1500(s, c, om);
    let b = (SR_PRESSURE_1500_KPA.ln() - SR_PRESSURE_33_KPA.ln()) / (t33.ln() - t1500.ln());
    1.0 / b
}

fn sr_ksat(s: f64, c: f64, om: f64) -> f64 {
    let ts = sr_theta_s(s, c, om);
    let t33 = sr_theta_33(s, c, om);
    let lam = sr_lambda(s, c, om);
    SR_KSAT_COEFF * (ts - t33).powf(SR_KSAT_EXPONENT - lam)
}

/// Compute all Saxton-Rawls hydraulic properties from soil texture and OM.
///
/// # Arguments
/// * `input` — Sand/clay fractions (0–1) and organic matter percentage.
///
/// # Returns
/// Full set of hydraulic properties including `θ_wp`, `θ_fc`, `θ_s`, Ksat, and λ.
///
/// # Reference
/// Saxton KE, Rawls WJ (2006) Soil Sci. Soc. Am. J. 70(5):1569-1578.
#[must_use]
pub fn saxton_rawls(input: &SaxtonRawlsInput) -> SaxtonRawlsResult {
    let (s, c, om) = (input.sand, input.clay, input.om_pct);
    SaxtonRawlsResult {
        theta_wp: sr_theta_1500(s, c, om),
        theta_fc: sr_theta_33(s, c, om),
        theta_s: sr_theta_s(s, c, om),
        ksat_mm_hr: sr_ksat(s, c, om),
        lambda: sr_lambda(s, c, om),
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used, reason = "test code uses unwrap for clarity")]
#[allow(clippy::expect_used, reason = "test code may use expect")]
mod tests {
    use super::*;

    #[test]
    fn loam() {
        let input = SaxtonRawlsInput {
            sand: 0.40,
            clay: 0.20,
            om_pct: 2.5,
        };
        let r = saxton_rawls(&input);
        assert!(r.theta_wp > 0.08 && r.theta_wp < 0.20, "wp={}", r.theta_wp);
        assert!(r.theta_fc > 0.20 && r.theta_fc < 0.40, "fc={}", r.theta_fc);
        assert!(r.theta_s > 0.35 && r.theta_s < 0.55, "θs={}", r.theta_s);
        assert!(
            r.ksat_mm_hr > 1.0 && r.ksat_mm_hr < 100.0,
            "Ksat={}",
            r.ksat_mm_hr
        );
        assert!(r.theta_wp < r.theta_fc, "wp < fc");
        assert!(r.theta_fc < r.theta_s, "fc < θs");
    }

    #[test]
    fn sand() {
        let input = SaxtonRawlsInput {
            sand: 0.92,
            clay: 0.03,
            om_pct: 1.0,
        };
        let r = saxton_rawls(&input);
        assert!(
            r.ksat_mm_hr > 50.0,
            "Sand Ksat should be high: {}",
            r.ksat_mm_hr
        );
        assert!(r.theta_wp < 0.10, "Sand WP should be low: {}", r.theta_wp);
    }

    #[test]
    fn clay() {
        let input = SaxtonRawlsInput {
            sand: 0.20,
            clay: 0.55,
            om_pct: 2.0,
        };
        let r = saxton_rawls(&input);
        assert!(
            r.ksat_mm_hr < 10.0,
            "Clay Ksat should be low: {}",
            r.ksat_mm_hr
        );
        assert!(r.theta_wp > 0.15, "Clay WP should be high: {}", r.theta_wp);
    }

    #[test]
    fn ordering() {
        let sand = saxton_rawls(&SaxtonRawlsInput {
            sand: 0.92,
            clay: 0.03,
            om_pct: 1.0,
        });
        let clay = saxton_rawls(&SaxtonRawlsInput {
            sand: 0.20,
            clay: 0.55,
            om_pct: 2.0,
        });
        assert!(sand.ksat_mm_hr > clay.ksat_mm_hr);
        assert!(clay.theta_wp > sand.theta_wp);
        assert!(clay.theta_fc > sand.theta_fc);
    }

    #[test]
    fn om_sensitivity() {
        let lo = saxton_rawls(&SaxtonRawlsInput {
            sand: 0.40,
            clay: 0.20,
            om_pct: 0.5,
        });
        let hi = saxton_rawls(&SaxtonRawlsInput {
            sand: 0.40,
            clay: 0.20,
            om_pct: 5.0,
        });
        assert!(hi.theta_wp > lo.theta_wp, "Higher OM → higher WP");
        assert!(hi.theta_fc > lo.theta_fc, "Higher OM → higher FC");
    }
}
